config/configoptional/optional_test.go

```diff
@@ -365,3 +367,94 @@ func TestComparePointer(t *testing.T) {
 		})
 	}
 }
+
+func TestOptionalMarshal(t *testing.T) {
+	tests := []struct {
+		name     string
+		value    Config[Sub]
+		expected map[string]any
+	}{
+		{
+			name:     "none (zero value)",
+			value:    Config[Sub]{},
+			expected: map[string]any{"sub": nil},
+		},
+		{
+			name:     "none",
+			value:    Config[Sub]{Sub1: None[Sub]()},
+			expected: map[string]any{"sub": nil},
+		},
+		{
+			name:     "default",
+			value:    Config[Sub]{Sub1: Default(subDefault)},
+			expected: map[string]any{"sub": nil},
+		},
+		{
+			name: "some",
+			value: Config[Sub]{Sub1: Some(Sub{
+				Foo: "bar",
+			})},
+			expected: map[string]any{
+				"sub": map[string]any{
+					"foo": "bar",
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			conf := confmap.New()
+			require.NoError(t, conf.Marshal(test.value))
+			assert.Equal(t, test.expected, conf.ToStringMap())
+		})
+	}
+}
+
+func TestComparePointerMarshal(t *testing.T) {
```

Comment:
Doesn't have to be added in this PR since this test covers it, but I think a similar test that verifies these properties with a generic Optional-style wrapper struct would be good in confmap since the changes in confmap essentially allow structs to declare themselves as automatically squashed when marshaling.

Reply:
> confmap essentially allow structs to declare themselves as automatically squashed when marshaling

I couldn't quite figure out exactly what is meant here. How does this happen?

Reply:
We discussed this on the Collector stability meeting on Monday June 16th and clarified that this is equivalent to adding a squash tag on the internal struct (which we can't do here exactly, but it serves as a useful analogy)